pkg/test/fakes/imageregistry/main.go

```diff
@@ -0,0 +1,113 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/tls"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"istio.io/pkg/log"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	containerregistry "github.com/google/go-containerregistry/pkg/registry"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+)
+
+var port = flag.Int("port", 1338, "port to run registry on")
+var images = flag.String("images", "", "comma separated list of images that will be preloaded and served")
+var registry = flag.String("registry", "localhost:1338", "name of registry to push and pull image")
+
+const (
+	User   = "user"
+	Passwd = "passwd"
+)
+
+type Handler struct {
+	registry http.Handler
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	encoded := base64.StdEncoding.EncodeToString([]byte("user:passwd"))
+	authHdr := r.Header.Get("Authorization")
+	wantHdr := fmt.Sprintf("Basic %s", encoded)
+	if authHdr != wantHdr {
+		log.Infof("Unauthorized: " + r.URL.Path)
+		log.Infof("Got header %v want header %v", authHdr, wantHdr)
+		w.Header().Set("WWW-Authenticate", "Basic")
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+	h.registry.ServeHTTP(w, r)
+}
+
+func main() {
+	flag.Parse()
+	s := &http.Server{
```

Comment:
do we need support SSL?

Reply:
I think it is fine to serve and test with only plaintext? SSL or not SSL is not the really relevant to Wasm image pulling.

Reply:
I don't think we need integration test to verify that? Should be easy to just have a unit test with a test server?

Reply:
Yeah same feeling. If the concern is whether env variable can be applied to proxy via ProxyMetadata, that path should already well covered by other tests.